src/icon.rs

```diff
@@ -1,353 +1,377 @@
-        } else if let FileType::Socket = file_type {
-            "\u{f6a7}" // ""
-        } else if let FileType::Pipe = file_type {
-            "\u{f731}" // ""
-        } else if let FileType::CharDevice = file_type {
-            "\u{e601}" // ""
-        } else if let FileType::BlockDevice = file_type {
-            "\u{fc29}" // "ﰩ"
-        } else if let FileType::Special = file_type {
-            "\u{f2dc}" // ""
-        } else if let Some(icon) = self
-            .icons_by_name
-            .get(name.file_name().to_lowercase().as_str())
-        {
-            // Use the known names.
-            icon
-        } else if let Some(icon) = name.extension().and_then(|extension| {
-            self.icons_by_extension
-                .get(extension.to_lowercase().as_str())
-        }) {
-            // Use the known extensions.
-            icon
-        } else {
-            // Use the default icons.
-            self.default_file_icon
-        };
+        match name.file_type() {
+            FileType::Directory { .. } => Some(self.default_folder_icon),
+            FileType::SymLink { is_dir: true } => Some('\u{f482}'), // ""
+            FileType::SymLink { is_dir: false } => Some('\u{f481}'), // ""
+            FileType::Socket => Some('\u{f6a7}'),                   // ""
+            FileType::Pipe => Some('\u{f731}'),                     // ""
+            FileType::CharDevice => Some('\u{e601}'),               // ""
+            FileType::BlockDevice => Some('\u{fc29}'),              // "ﰩ"
+            FileType::Special => Some('\u{f2dc}'),                  // ""
+            FileType::File { .. } => self
+                .icons_by_name
+                .get(name.get_name().to_ascii_lowercase().as_str())
+                .or_else(|| {
+                    if let Some(ext) = name.extension() {
+                        return self
+                            .icons_by_extension
+                            .get(ext.to_ascii_lowercase().as_str());
+                    }
+                    let mut reader = BufReader::new(File::open(name.get_path()).ok()?);
+                    let mut buf = [0; 2];
+                    reader.read_exact(&mut buf).ok()?;
+                    if b"#!" != &buf {
+                        return None;
+                    }
+                    let line = reader.lines().next()?.ok()?;
+                    let end_path = line.split('/').last()?;
+                    let command = if end_path.starts_with("env") {
+                        end_path.split(' ').next_back() // #!/bin/env bash
+                    } else {
+                        end_path.split(' ').next() // #!/bin/bash -vv
+                    }?;
+                    self.icons_by_shebang(command)
+                })
+                .cloned()
+                .or(Some(self.default_file_icon)),
+        }
+    }
 
-        format!("{}{}", icon, self.icon_separator)
+    fn icons_by_shebang(&self, cmd: &str) -> Option<&char> {
```

Comment:
This is a great addition.

Reply:
thanks! i  got tired of my bin folders being blank files :laughing: